### PR TITLE
Added arguments to fakefish + customizable tls settings

### PR DESCRIPTION
**Pull request by mvazquezc**
Added arguments to the tool. On top of that made it possible to use different configs for the TLS mode.

### CODE REVIEW

File: app/fakefish.py [hunk 1]
```diff
@@ -172,21 +164,36 @@ def set_env_vars(bmc_endpoint, username, password):
...
     """
 
     """
-    app.run(host='0.0.0.0', port=port, debug=debug, ssl_context='adhoc')
+    if tls_mode == 'adhoc':
+        app.run(host='0.0.0.0', port=port, debug=debug, ssl_context='adhoc')
+    elif tls_mode == 'disabled':
+        app.run(host='0.0.0.0', port=port, debug=debug)
+    else:
+        if os.path.exists('./cert.pem') and os.path.exists('./cert.key'):
```
Comment (dtantsur) [inline app/fakefish.py:176]: Ideally, these should be parameters.

File: app/fakefish.py [hunk 2]
```diff
@@ -172,21 +164,36 @@ def set_env_vars(bmc_endpoint, username, password):
...
 
 
 if __name__ == '__main__':
-
-    bmc_ip = os.environ.get('BMC_IP', None)
-    if bmc_ip is None:
-        app.logger.error('Configure the BMC IP using the environment variable BMC_IP')
-        exit()
+    parser = argparse.ArgumentParser(description='FakeFish, an experimental RedFish proxy that calls shell scripts for executing hardware actions.')
+    parser.add_argument('--tls-mode', type=str, choices=['adhoc', 'self-signed', 'disabled'], default='adhoc', help='Configures TLS mode. \'self-signed\' mode expects cert.pem and cert.key files within the app path. (default: %(default)s)')
```
Comment (dtantsur) [inline app/fakefish.py:185]: Could you please keep the code wrapped at column 80?

File: README.md [hunk 3]
```diff
@@ -39,14 +41,23 @@ A [Containerfile](./custom_scripts/Containerfile) is included, so users can buil
...
 You need a FakeFish process for each server you plan to install. Think of FakeFish like if it was a custom implementation of a BMC for that specific server.
 
-Since you will be potentially running multiple FakeFish instances, you will make use of an environment variable to configure in which port a given FakeFish instance listens on and another one to configure which BMC IP it provides service to. On top of that, you can do a bind mount for the folder containing the scripts for managing that specific server or use the scripts inside the container image.
+Currently, FakeFish exposes the following arguments:
+
+|Argument|What is it used for|Default|Required|
+|--------|-------------------|-------|--------|
+|`--tls-mode`| Configures TLS for FakeFish, three available modes: `adhoc` (FakeFish generated certs), `self-signed` (user provided certs), `disabled` (TLS disabled).|`adhoc`|No|
+|`--managed-endpoint` or `-e`|Defines the BMC IP a FakeFish instance will connect to.|`None`|Yes|
+|`--port` or `-p`|Defines the port a FakeFish instance will listen on.|`9000`|No|
```
Comment (dtantsur) [inline README.md:50]: Maybe `--listen-port` to avoid confusion with the BMC port.

File: app/fakefish.py [hunk 4]
```diff
@@ -172,21 +164,36 @@ def set_env_vars(bmc_endpoint, username, password):
...
     return my_env
 
-def run():
+def run(port, debug, tls_mode):
     """
 
     """
-    app.run(host='0.0.0.0', port=port, debug=debug, ssl_context='adhoc')
+    if tls_mode == 'adhoc':
+        app.run(host='0.0.0.0', port=port, debug=debug, ssl_context='adhoc')
```
Comment (dtantsur) [inline app/fakefish.py:172]: Side comment: '0.0.0.0' is not IPv6-friendly, I assume. Maybe `::`

File: app/fakefish.py (hunk 2 above)
Comment (mvazquezc) [reply app/fakefish.py:185]: Done.

File: README.md [hunk 5]
```diff
@@ -39,14 +41,23 @@ A [Containerfile](./custom_scripts/Containerfile) is included, so users can buil
 
 You need a FakeFish process for each server you plan to install. Think of FakeFish like if it was a custom implementation of a BMC for that specific server.
 
-Since you will be potentially running multiple FakeFish instances, you will make use of an environment variable to configure in which port a given FakeFish instance listens on and another one to configure which BMC IP it provides service to. On top of that, you can do a bind mount for the folder containing the scripts for managing that specific server or use the scripts inside the container image.
+Currently, FakeFish exposes the following arguments:
+
+|Argument|What is it used for|Default|Required|
+|--------|-------------------|-------|--------|
+|`--tls-mode`| Configures TLS for FakeFish, three available modes: `adhoc` (FakeFish generated certs), `self-signed` (user provided certs), `disabled` (TLS disabled).|`adhoc`|No|
+|`--managed-endpoint` or `-e`|Defines the BMC IP a FakeFish instance will connect to.|`None`|Yes|
```
[the remark this answers is not in the document]
Comment (mvazquezc) [reply README.md:49]: Done.

File: README.md (hunk 3 above)
Comment (mvazquezc) [reply README.md:50]: Done.

File: app/fakefish.py (hunk 4 above)
Comment (mvazquezc) [reply app/fakefish.py:172]: Done.

File: dell_scripts/Containerfile [hunk 6]
```diff
@@ -11,13 +11,16 @@ RUN set -x && \
 
 RUN mkdir -p /opt/fakefish/
 
-COPY app/fakefish.py /opt/fakefish/fakefish.py
+# The stars in the command below will only copy those files if they exist
+COPY app/fakefish.py app/cert.pem* app/cert.key* /opt/fakefish/
```
[the remark this answers is not in the document]
Comment (mvazquezc) [reply dell_scripts/Containerfile:15]: Yes. Will do this in another PR.

**Comment by mvazquezc** [comment]
@dtantsur applied suggested changes.

File: README.md [hunk 7]
```diff
@@ -39,14 +41,25 @@ A [Containerfile](./custom_scripts/Containerfile) is included, so users can buil
...
+
+Since you will be potentially running multiple FakeFish instances, you will make use of the `--port` argument to configure in which port a given FakeFish instance listens on and `--managed-endpoint` to configure which BMC IP it provides service to. On top of that, you can do a bind mount for the folder containing the scripts for managing that specific server or use the scripts inside the container image.
 
 An example can be found below:
 
 > **NOTE**: Every container is mapped to a single BMC, but if more hosts are required, different ports can be used (9001, 9002,...)
 
 ~~~sh
-podman run -p 9000:9000 -e PORT=9000 -e BMC_IP=172.20.10.10 -v $PWD/dell_scripts:/opt/fakefish/custom_scripts:z quay.io/mavazque/fakefish:v0
+podman run -p 9000:9000 -v $PWD/dell_scripts:/opt/fakefish/custom_scripts:z quay.io/mavazque/fakefish:latest --port 9000 --managed-endpoint 172.20.10.10
```
Comment (dtantsur) [inline README.md:62]: Wrong argument, needs updating

File: README.md [hunk 8]
```diff
@@ -64,7 +77,7 @@ In a successful execution you should see something like this in the logs:
 - Starting FakeFish
 
     ~~~sh
-    $ podman run -p 9000:9000 -e PORT=9000 -e BMC_IP=172.20.10.10 -v $PWD/dell_scripts:/opt/fakefish/custom_scripts:z quay.io/mavazque/fakefish:v0
+    $ podman run -p 9000:9000 -v $PWD/dell_scripts:/opt/fakefish/custom_scripts:z quay.io/mavazque/fakefish:latest --port 9000 --managed-endpoint 172.20.10.10
```
Comment (dtantsur) [inline README.md:80]: Same

File: app/fakefish.py [hunk 9]
```diff
@@ -172,21 +164,41 @@ def set_env_vars(bmc_endpoint, username, password):
...
-        app.logger.error('Configure the BMC IP using the environment variable BMC_IP')
-        exit()
+    parser = argparse.ArgumentParser(description='FakeFish, an experimental RedFish proxy that calls shell scripts for executing hardware actions.')
+    parser.add_argument('--tls-mode', type=str, choices=['adhoc', 'self-signed', 'disabled'], default='adhoc', help='Configures TLS mode. \
+                        \'self-signed\' mode expects users to configure a cert and a key files. (default: %(default)s)')
+    parser.add_argument('--cert-file', type=str, default='./cert.pem', help='Path to the certificate public key file. (default: %(default)s)')
+    parser.add_argument('--key-file', type=str, default='./cert.key', help='Path to the certificate private key file. (default: %(default)s)')
+    parser.add_argument('-r', '--remote-bmc', type=str, required=True, help='The BMC IP this FakeFish instance will connect to. e.g: 192.168.1.10')
+    parser.add_argument('-p','--listen-port', type=int, required=False, default=9000, help='The port where this FakeFish instance will listen for connections.')
+    parser.add_argument('-d', '--debug', action='store_true')
```
Comment (dtantsur) [inline app/fakefish.py:191]: nit: I tend to avoid a short variant for --debug since it shouldn't normally by used

File: README.md (hunk 7 above)
Comment (mvazquezc) [reply README.md]: Fixed

File: app/fakefish.py (hunk 9 above)
Comment (mvazquezc) [reply app/fakefish.py]: Changed.

**Pull request merged**